packages/block-library/src/navigation/edit/index.js

```diff
@@ -143,6 +143,8 @@ function Navigation( {
 		);
 	}
 
+	const isCreatingEmpty = useRef( false );
```

Comment:
If I'm understanding properly this ref is only needed because the user can create empty menus explicitly before the navigation menus load right? 

Can't we just hide all UI (keep the loader) while the navigation menus are being loaded?

Reply:
We need it because when the user clicks `Create empty` it clears out any existing ref`. At this point the block considers itself "empty" and thus tries to automatically utilise the first Navigation Menu if it exists. This means you can never get an empty menu cos it's always auto-filled with the first Nav Menu.

The flag exists simply to block that happening if the user has explicitly said "I want a blank Nav please". 

There are two states here:

1. Empty block that is inserted into canvas or already exists in canvas.
2. An empty menu that the user has created.

> Can't we just hide all UI (keep the loader) while the navigation menus are being loaded?

I tried that yes. It means the block takes longer for the Nav to reach a "settled" state. Moreover, that doesn't solve the issue I mentioned above because even if the block auto selects the first Nav Menu, sometimes as a user you want a entirely blank menu. Therefore you would click `Create empty` and because the block would already have the Nav Menu data available it would jump immediately to populating with the first one.

Reply:
Is it possible to infer that state (wanted empty menu) from state instead of using a ref?

The current implementation is a bit fragile and relying on the promise returned by data module actions in event callbacks is not a good practice. (React warning if the menu is removed before the action finishes)

Reply:
Thanks for following up here.

> The current implementation is a bit fragile and relying on the promise returned by data module actions in event callbacks is not a good practice.

It's certainly sub optimal. I can take a look again at whether we can infer from state but I seem to remember that wasn't possible.

The question I have is whether this is "good enough" (short term) to fix the problem which results in menus being created on focusing the block (which is pretty disruptive).

We can always follow up with a revised implementation post-6.0. But if you feel this is sufficiently brittle let's call it now and focus our efforts elsewhere. 

In the meantime I'll dig in again.


> React warning if the menu is removed before the action finishes

Yeh I've discussed this a few times with folks. The [last opinion given by @adamziel](https://github.com/WordPress/gutenberg/pull/38956#issuecomment-1047744390) was that this warning is being removed in React now anyway so let's not worry about it. I assume we're concerned about potential memory leaks?

Reply:
> The current implementation is a bit fragile and relying on the promise returned by data module actions in event callbacks is not a good practice.

@youknowriad I removed the reliance on `.finally()` and instead utilised the statuses I created in a previous PR to reset the flag.

I'd also like to understand more about why relying on the returned `Promise` is not considered to be a good practice.

Reply:
@getdave I think that `isCreatingEmpty.current` could be replaced with `isSavingEntityRecord( 'postType', 'wp_navigation' )` without an id argument. The latter "Returns true if the specified entity record is saving, and false otherwise."

Reply:
Hm although the ref is specific to the block instance, whereas `isSavingEntityRecord` would flip true/false in all the blocks in the editor.

Reply:
Remember we have to detect whether the _user_ is creating an _empty_ menu. It's not just "we're creating a menu".

Reply:
Ok I've moved to using existing state rather than `useRef()`. Thanks for flagging this. It was a red flag there there might have been a simpler solution.